File: android/sdk/src/main/java/org/jitsi/meet/sdk/VideoConfBridge.java
/*
 * Copyright © 2020. cabi LLC. All rights reserved.
 */

package org.jitsi.meet.sdk;

import org.jitsi.meet.sdk.ReactInstanceManagerHolder;

public class VideoConfBridge {

    /**
     * Videoconf join event.
     */
    private static final String VIDEOCONF_JOIN = "org.jitsi.meet:features/videoconf-bridge#join-conference";

    /**
     * Videoconf leave event.
     */
    private static final String VIDEOCONF_LEAVE = "org.jitsi.meet:features/videoconf-bridge#leave-conference";

    /**
     * Media mute/unmute event.
     */
    private static final String MUTE_MEDIA = "org.jitsi.meet:features/videoconf-bridge#mute-media";

    /**
     * Switch camera.
     */
    private static final String SWITCH_CAMERA = "org.jitsi.meet:features/videoconf-bridge#switch-camera";

    /**
     * Send Jitsi command.
     */
    private static final String SEND_COMMAND = "org.jitsi.meet:features/videoconf-bridge#send-command";

    /**
     * Remove Jitsi command.
     */
    private static final String REMOVE_COMMAND = "org.jitsi.meet:features/videoconf-bridge#remove-command";

    /**
     * Set current swiper index.
     */
    private static final String SET_CURRENT_SWIPER_INDEX = "org.jitsi.meet:features/videoconf-bridge#set-current-swiper-index";

    /**
     * Show wrap up buttons in TileView.
     */
    private static final String SHOW_WRAP_UP_BUTTONS = "org.jitsi.meet:features/videoconf-bridge#show-wrap-up-buttons";

    /**
     * In-focus placeholder data.
     */
    private static final String PLACEHOLDER_DATA = "org.jitsi.meet:features/videoconf-bridge#placeholder-data";

    /**
     * Set pre show countdown date time
     */
    private static final String SET_COUNTDOWN = "org.jitsi.meet:features/videoconf-bridge#set-countdown";

    /**
     * View change event.
     */
    private static final String SHOW_SPEAKER_VIEW = "org.jitsi.meet:features/videoconf-bridge#show-speaker-view";

    /**
     * Update video conference user avatar event.
     */
    private static final String UPDATE_USER_AVATAR_EVENT = "org.jitsi.meet:features/videoconf-bridge#update-user-avatar";

    /**
     * Event for muting/unmuting remote audio tracks in video conference locally.
     */
    private static final String MUTE_CONFERENCE_AUDIO = "org.jitsi.meet:features/videoconf-bridge#mute-conference-audio";

    /**
     * Sets whether video conference is simplified.
     */
    private static final String SET_IS_SIMPLIFIED_CONFERENCE = "org.jitsi.meet:features/videoconf-bridge#set-is-simplified-conference";

    /**
     * Sends a videoconf join event to React Native.
     */
    public static void joinConference(String dataJsonString) {
        ReactInstanceManagerHolder.emitEvent(VIDEOCONF_JOIN, dataJsonString);
    }

    /**
     * Sends a videoconf leave event to React Native.
     */
    public static void leaveConference() {
        ReactInstanceManagerHolder.emitEvent(VIDEOCONF_LEAVE, null);
    }

    /**
     * Mute/unmute mic/cam described by stringified data json parameter:
     * {
     *   kind: 'audioinput' or 'videoinput',
     *   muted: boolean
     * }
     */
    public static void muteMedia(String dataJsonString) {
        ReactInstanceManagerHolder.emitEvent(MUTE_MEDIA, dataJsonString);
    }

    /**
     * Switch camera between front-back.
     */
    public static void switchCamera() {
        ReactInstanceManagerHolder.emitEvent(SWITCH_CAMERA, null);
    }

    /**
     * Send jitsi command in confernce.
     */
    public static void sendJitsiCommand(String dataJsonString) {
        ReactInstanceManagerHolder.emitEvent(SEND_COMMAND, dataJsonString);
    }

    /**
     * Remove jitsi command from presence.
     */
    public static void removeJitsiCommand(String commandName) {
        ReactInstanceManagerHolder.emitEvent(REMOVE_COMMAND, commandName);
    }

    /**
     * Swipe to given page.
     */
    public static void setCurrentSwiperIndex(String pageNumber) {
        ReactInstanceManagerHolder.emitEvent(SET_CURRENT_SWIPER_INDEX, pageNumber);
    }

    /**
     * Show wrap up buttons.
     */
    public static void showWrapUpButtons() {
        ReactInstanceManagerHolder.emitEvent(SHOW_WRAP_UP_BUTTONS, null);
    }

    /**
     * Sends in-focus placeholder data to React Native.
     */
    public static void sendPlaceholderData(String data) {
        ReactInstanceManagerHolder.emitEvent(PLACEHOLDER_DATA, data);
    }

    /**
     * Set pre show countdown date time.
     */
    public static void setCountdown(String datetime) {
        ReactInstanceManagerHolder.emitEvent(SET_COUNTDOWN, datetime);
    }

    /**
     * Update video conference user profile image.
     */
    public static void updateUserAvatar(String data) {
        ReactInstanceManagerHolder.emitEvent(UPDATE_USER_AVATAR_EVENT, data);
    }

    /**
     * Sends a view change event to React Native.
     */
    public static void showSpeakerView(boolean shouldShow) {
        ReactInstanceManagerHolder.emitEvent(SHOW_SPEAKER_VIEW, shouldShow);
    }

    /**
     * Mute/unmute remote audio tracks in video conference locally.
     */
    public static void muteVideoConferenceAudio(boolean mute) {
        ReactInstanceManagerHolder.emitEvent(MUTE_CONFERENCE_AUDIO, mute);
    }

    /**
     * Sets whether video conference is simplified.
     */
    public static void setIsSimplifiedConference(final boolean isSimplifiedConference) {
        ReactInstanceManagerHolder.emitEvent(SET_IS_SIMPLIFIED_CONFERENCE, isSimplifiedConference);
    }
}

File: ios/sdk/src/VideoConfBridge.m
//
//  VideoConfBridge.m
//  JitsiMeet
//
//  Created by Molnár Tímea on 2020. 10. 29..
//  Copyright © 2020. cabi LLC. All rights reserved.
//

#import <Foundation/Foundation.h>

#import "VideoConfBridge.h"
#import "JitsiMeetEventEmitter+Private.h"

@implementation VideoConfBridge

static NSString *const JOIN_EVENT = @"org.jitsi.meet:features/videoconf-bridge#join-conference";
static NSString *const LEAVE_EVENT = @"org.jitsi.meet:features/videoconf-bridge#leave-conference";
static NSString *const MUTE_EVENT = @"org.jitsi.meet:features/videoconf-bridge#mute-media";
static NSString *const SWITCH_CAM_EVENT = @"org.jitsi.meet:features/videoconf-bridge#switch-camera";
static NSString *const SEND_COMMAND_EVENT = @"org.jitsi.meet:features/videoconf-bridge#send-command";
static NSString *const REMOVE_COMMAND_EVENT = @"org.jitsi.meet:features/videoconf-bridge#remove-command";
static NSString *const SWIPE_TO_PAGE_EVENT = @"org.jitsi.meet:features/videoconf-bridge#set-current-swiper-index";
static NSString *const SHOW_WRAP_UP_BUTTONS = @"org.jitsi.meet:features/videoconf-bridge#show-wrap-up-buttons";
static NSString *const PLACEHOLDER_DATA_EVENT = @"org.jitsi.meet:features/videoconf-bridge#placeholder-data";
static NSString *const SET_COUNTDOWN_EVENT = @"org.jitsi.meet:features/videoconf-bridge#set-countdown";
static NSString *const SHOW_SPEAKER_VIEW_EVENT = @"org.jitsi.meet:features/videoconf-bridge#show-speaker-view";
static NSString *const UPDATE_USER_AVATAR_EVENT = @"org.jitsi.meet:features/videoconf-bridge#update-user-avatar";
static NSString *const MUTE_CONFERENCE_AUDIO = @"org.jitsi.meet:features/videoconf-bridge#mute-conference-audio";
static NSString *const SET_IS_SIMPLIFIED_CONFERENCE = @"org.jitsi.meet:features/videoconf-bridge#set-is-simplified-conference";


RCT_EXPORT_MODULE();

- (void)registerSelfToEventEmitter {
    [[JitsiMeetEventEmitter sharedEmitter] registerVideoConfBridge:self];
}


- (NSArray<NSString *> *)supportedEvents {
  return @[JOIN_EVENT,
        LEAVE_EVENT,
        MUTE_EVENT,
        SWITCH_CAM_EVENT,
        SEND_COMMAND_EVENT,
        REMOVE_COMMAND_EVENT,
        PLACEHOLDER_DATA_EVENT,
        SWIPE_TO_PAGE_EVENT,
        SET_COUNTDOWN_EVENT,
        SHOW_SPEAKER_VIEW_EVENT,
        UPDATE_USER_AVATAR_EVENT,
        MUTE_CONFERENCE_AUDIO,
        SET_IS_SIMPLIFIED_CONFERENCE];
}

- (void)join:(NSString *)dataJsonString {
    [self sendEvent:JOIN_EVENT body:dataJsonString];
}

- (void)leave {
    [self sendEvent:LEAVE_EVENT body:nil];
}

- (void)muteMedia:(NSString *_Nonnull)dataJsonString {
    [self sendEvent:MUTE_EVENT body:dataJsonString];
}

- (void)switchCamera {
    [self sendEvent:SWITCH_CAM_EVENT body:nil];
}

- (void)sendCommand:(NSString *_Nonnull)dataJsonString {
    [self sendEvent:SEND_COMMAND_EVENT body:dataJsonString];
}

- (void)removeCommand:(NSString *_Nonnull)commandName {
    [self sendEvent:REMOVE_COMMAND_EVENT body:commandName];
}

- (void)sendPlaceholderData:(NSString *_Nonnull)data {
    [self sendEvent:PLACEHOLDER_DATA_EVENT body:data];
}

- (void)setCurrentSwiperIndex:(NSString *_Nonnull)pageNumber {
    [self sendEvent:SWIPE_TO_PAGE_EVENT body:pageNumber];
}

- (void)showWrapUpButtons {
    [self sendEvent:SHOW_WRAP_UP_BUTTONS body:nil];
}

- (void)setCountdown:(NSString *_Nonnull)jsonString {
    [self sendEvent:SET_COUNTDOWN_EVENT body:jsonString];
}

- (void)showSpeakerView:(NSNumber *_Nonnull)show {
    [self sendEvent:SHOW_SPEAKER_VIEW_EVENT body:show];
}

- (void)updateUserAvatar:(NSString *_Nonnull) jsonString {
    [self sendEvent:UPDATE_USER_AVATAR_EVENT body:jsonString];
}

- (void)muteVideoConferenceAudio:(NSNumber *_Nonnull)mute {
    [self sendEvent:MUTE_CONFERENCE_AUDIO body:mute];
}

- (void)setIsSimplifiedConference:(NSNumber *_Nonnull)isSimplifiedConference {
    [self sendEvent:SET_IS_SIMPLIFIED_CONFERENCE body:isSimplifiedConference];
}

@end
